src/contract.rs

```diff
@@ -27,45 +28,70 @@ pub struct Auction {
     pub status: Status,
     pub started_at: u64,
     pub expires_at: u64,
+
+    pub transactions: BTreeMap<ActorId, Transaction<Action>>,
+    pub current_tid: TransactionId,
 }
 
 impl Auction {
-    pub async fn buy(&mut self) {
-        if !self.is_active() {
-            panic!("already bought or auction expired");
+    pub async fn buy(&mut self, transaction_id: TransactionId) -> Result<(Event, u128), Error> {
+        if !matches!(self.status, Status::IsRunning) {
+            return Err(Error::AlreadyStopped);
         }
 
         if exec::block_timestamp() >= self.expires_at {
-            panic!("auction expired");
+            return Err(Error::Expired);
         }
 
         let price = self.token_price();
 
         if msg::value() < price {
-            panic!("value < price, {:?} < {:?}", msg::value(), price);
+            gstd::debug!("value < price, {:?} < {:?}", msg::value(), price);
+            return Err(Error::InsufficentMoney);
         }
 
         self.status = Status::Purchased { price };
 
         let refund = msg::value() - price;
         let refund = if refund < 500 { 0 } else { refund };
-        let transaction_id = 0u64;
 
-        msg::send_for_reply(
+        gstd::debug!("before Transfer NFT");
+
+        let reply = match msg::send_for_reply(
             self.nft.contract_id,
             NFTAction::Transfer {
                 to: msg::source(),
                 token_id: self.nft.token_id,
                 transaction_id,
             },
             0,
-        )
-        .unwrap()
-        .await
-        .expect("Error in nft transfer");
+        ) {
+            Ok(reply) => {
+                gstd::debug!("Send OK");
+                reply
+            }
+            Err(e) => {
+                gstd::debug!("Send Error {:?}", e);
+                return Err(Error::NftTransferFailed);
+            }
+        };
+
+        match reply.await {
+            Ok(_reply) => gstd::debug!("Reply Ok"),
+            Err(e) => {
+                gstd::debug!("Await Reply Error {:?}", e);
+                return Err(Error::NftTransferFailed);
+            }
+        }
+
+        gstd::debug!("before Transfer Reward");
+
+        if let Err(e) = msg::send(self.nft.owner, "REWARD", price) {
```

Comment:
What if NFT is transferred and the gas ran out here?
The value is already on the contract
And the buyer may not continue the transaction after receiving the NFT